contracts/optimistic-ethereum/OVM/verification/OVM_StateTransitioner.sol

```diff
@@ -173,81 +182,69 @@ contract OVM_StateTransitioner is Lib_AddressResolver, OVM_FraudContributor, iOV
         require(
-            ovmStateManager.hasAccount(_ovmContractAddress) == false,
-            "Account state has already been proven"
-        );
-
-        require(
-            _account.codeHash == Lib_EthUtils.getCodeHash(_ethContractAddress),
-            "Invalid code hash provided."
-        );
-
-        require(
-            Lib_SecureMerkleTrie.verifyInclusionProof(
-                abi.encodePacked(_ovmContractAddress),
-                Lib_OVMCodec.encodeEVMAccount(_account),
-                _stateTrieWitness,
-                preStateRoot
+            (
+                ovmStateManager.hasAccount(_ovmContractAddress) == false
+                && ovmStateManager.hasEmptyAccount(_ovmContractAddress) == false
             ),
-            "Account state is not correct or invalid inclusion proof provided."
-        );
-
-        ovmStateManager.putAccount(
-            _ovmContractAddress,
-            Lib_OVMCodec.Account({
-                nonce: _account.nonce,
-                balance: _account.balance,
-                storageRoot: _account.storageRoot,
-                codeHash: _account.codeHash,
-                ethAddress: _ethContractAddress,
-                isFresh: false
-            })
-        );
-    }
-
-    /**
-     * Allows a user to prove that an account does *not* exist in the state.
-     * @param _ovmContractAddress Address of the contract on the OVM.
-     * @param _stateTrieWitness Proof of the (empty) account state.
-     */
-    function proveEmptyContractState(
-        address _ovmContractAddress,
-        bytes memory _stateTrieWitness
-    )
-        override
-        public
-        onlyDuringPhase(TransitionPhase.PRE_EXECUTION)
-        contributesToFraudProof(preStateRoot, transactionHash)
-    {
-        // Exit quickly to avoid unnecessary work.
-        require(
-            ovmStateManager.hasEmptyAccount(_ovmContractAddress) == false,
             "Account state has already been proven."
         );
 
-        require(
-            Lib_SecureMerkleTrie.verifyExclusionProof(
-                abi.encodePacked(_ovmContractAddress),
-                _stateTrieWitness,
-                preStateRoot
-            ),
-            "Account is not empty or invalid inclusion proof provided."
+        // Function will fail if the proof is invalid.
+        (
+            bool exists,
+            bytes memory encodedAccount
+        ) = Lib_SecureMerkleTrie.get(
+            abi.encodePacked(_ovmContractAddress),
+            _stateTrieWitness,
+            preStateRoot
         );
 
-        ovmStateManager.putEmptyAccount(_ovmContractAddress);
+        if (exists == true) {
+            // Account exists, this was an inclusion proof.
+            Lib_OVMCodec.EVMAccount memory account = Lib_OVMCodec.decodeEVMAccount(
+                encodedAccount
+            );
+
+            address ethContractAddress = _ethContractAddress;
+            if (account.codeHash == EMPTY_ACCOUNT_CODE_HASH) {
```

Comment:
I'm actually a little confused by this -- can you clarify the exact scenario that will trigger this?  Intuitively, I see no way for there to be an inclusion proof in the OVM state that has empty code at the bottom, since we do not have EOAs and do not have `ovmSELFDESTRUCT`.

If there's no scenario that will trigger this but we just thought of it, still could be worth leaving in as it was a great catch--moreso just curious.  Can initcode return 0 bytes maybe?

Reply:
I believe the only scenario that can trigger this is when initcode returns 0 bytes (just tested on remix, definitely possible).